### PR TITLE
fix(openai): replay assistant text history as valid string-content input in Responses adapters

### LINKED ISSUE #3388: OpenAIModel (Responses, TypeScript) sends invalid assistant history shape, rejected by Bedrock Mantle
### SDK Language

TypeScript

### Steps to Reproduce

The OpenAI Responses adapter serializes assistant history into an invalid `input` item shape when replaying multi-turn conversations without server-side state (`stateful: false`).

For a two-turn conversation, `formatResponsesMessages` in `strands-ts/src/models/openai/responses-adapter.ts` emits the assistant turn as:

```json
{ "role": "assistant", "content": [{ "type": "output_text", "text": "..." }] }
```

This matches neither valid OpenAI `input` shape for an assistant turn:
- `EasyInputMessage` (string content): `{ "role": "assistant", "content": "..." }`
- `ResponseOutputMessage` (full item): requires `type: "message"`, `id`, `status`, and `output_text` with `annotations`.

The emitted object is a bare `{ role, content: [output_text] }` missing all the `ResponseOutputMessage` required fields, so it is not a valid input item. The code cast it away with `as unknown as ResponseInputItem`, and its own comment acknowledged the shape was incomplete.

Reproduce with the Bedrock Mantle integ test `test/integ/models/openai/mantle.test.node.ts > handles reasoning content across multi-turn conversations` (requires live Bedrock Mantle credentials).

### Expected Behavior

Text-only assistant history is serialized as a valid, OpenAI-documented shape and accepted by both OpenAI-hosted and Bedrock Mantle.

### Actual Behavior

Bedrock Mantle's strict Responses schema rejects the request:

```
SubmitRequestFailure: code=-32602, msg=220 validation errors for ResponsesRequest
```

OpenAI-hosted tolerates the malformed shape (fills in / ignores the missing fields), so the bug was invisible there. It only surfaced once PR #3290 made terminal `response.failed` stream events throw instead of finalizing as an empty successful response, which turned the previously-swallowed Mantle rejection into a loud failure.

### Possible Solution

N/A

### Related Issues

The Python SDK has the identical bug at `strands-py/src/strands/models/openai_responses.py` (`_format_request_message_content` emits `output_text` for assistant text); fix both for parity.

Surfaced by #3290.

### PR DESCRIPTION
**Pull request by agent-of-mkmeral**
## Description

Fixes the OpenAI Responses adapters (TS `OpenAIModel` in Responses mode and Python `OpenAIResponsesModel`) sending an invalid assistant history shape when replaying multi-turn conversations without server-side state.

### The bug

Assistant turns were serialized as:

```json
{ "role": "assistant", "content": [{ "type": "output_text", "text": "..." }] }
```

This matches **no** valid OpenAI `input` item shape:
- `EasyInputMessage` requires string content for assistant text (`input_text` in assistant messages is rejected — that was #1850, which this shape was the workaround for)
- `ResponseOutputMessage` requires `type: "message"`, `id`, `status`, and `annotations` — all missing

OpenAI-hosted silently tolerated the malformed item, but Bedrock Mantle's strict Responses schema rejects it (`SubmitRequestFailure: code=-32602, msg=220 validation errors for ResponsesRequest`). The TS code even force-cast it away with `as unknown as ResponseInputItem`.

### The fix

Text-only assistant history (including citation text) is now collapsed to the documented string-content `EasyInputMessage` — the only assistant *input* shape that is valid without the output-item metadata (`id`/`status`/`annotations`) the adapters have already discarded:

```json
{ "role": "assistant", "content": "..." }
```

- **TS** (`strands-ts/src/models/openai/responses-adapter.ts`): assistant content items are collapsed into a single string message at assembly time; the `as unknown as` force-cast on the assistant path is gone (the item now typechecks as a real `EasyInputMessage`, so `tsc` guards this shape again). Non-text assistant content (never a valid input shape) is warn-dropped. Drive-by: the `citationsBlock` case emitted `output_text` even for **user**-role messages — it is now role-aware (`input_text` for user), matching the Python side.
- **Python** (`strands-py/src/strands/models/openai_responses.py`): same collapse in `_format_request_messages`; `_format_request_message_content` keeps producing the `output_text` intermediate, which is now joined into the string form at final assembly. Non-text assistant content is warn-dropped.
- Multiple text blocks in one assistant message are joined with `\n` (same convention as text-only tool-result outputs).

Reasoning content is still dropped as before — proper verbatim output-item round-tripping (incl. `encrypted_content`) is tracked in #3389.

### Verification

**Unit / static:**
- Python: **104/104** tests in `test_openai_responses.py` pass (updated 4 tests that asserted the buggy shape; added 4 regression tests incl. one that validates the assistant item against the official `openai` SDK `EasyInputMessageParam` field contract). `ruff check` + `ruff format` clean.
- TS: **50/50** tests in `responses.test.ts` pass (3 new regression tests: string form, newline joining, non-text warn-drop); all 580 unit tests under `src/models/` pass; `tsc --noEmit` clean with the cast removed; `eslint`/`prettier` clean; full pre-commit hook (unit suite + coverage) passed.
- Schema strictness proxy for Mantle: validating the items against the official `openai` Python SDK `ResponseInputItemParam` union — **old shape: 115 validation errors** (mirrors Mantle's `220 validation errors for ResponsesRequest`); **new shape: valid**.

**Live E2E (OpenAI-hosted, real API key):**
- **Python**: two-turn `Agent` conversation via `OpenAIResponsesModel` (`gpt-4o-mini`); wire capture of the second request confirms the assistant history item is exactly `{"role": "assistant", "content": "2 + 2 equals 4."}`; both turns complete correctly.
- **TS**: same two-turn flow via `OpenAIModel` (Responses mode) with a request-capturing client; second request carries `{"role":"assistant","content":"2+2 equals 4."}`; both turns complete.
- **TS reasoning multi-turn** (mirrors the Mantle integ repro): `gpt-5-mini` with `reasoning: { effort: 'low' }`, two turns — second turn completes with `stopReason: endTurn` despite reasoning blocks in history.
- Raw-wire baseline check: OpenAI-hosted **accepts both** the old and new shapes, empirically confirming the issue's claim that the bug was invisible there and only surfaces on strict backends (Mantle).

**Not run locally:** the live Mantle integ test (`test/integ/models/openai/mantle.test.node.ts > handles reasoning content across multi-turn conversations`) — my credentials are denied `bedrock-mantle:CreateInference`, so the run failed with 401 access_denied before reaching the model. This is exactly what PR CI runs with team credentials; would appreciate confirmation from that run.

## Related Issues

Fixes #3388

## Type of Change

Bug fix

## Checklist
- [x] I have read the CONTRIBUTING document
- [x] My code follows the code style of this project
- [x] I have added tests that prove my fix is effective
- [x] All new and existing tests pass locally

### CODE REVIEW

**Comment by opieter-aws** [comment]
@agent-of-mkmeral check the codecov and failing integ tests. Review the comments.

File: strands-ts/src/models/openai/responses-adapter.ts [hunk 1]
```diff
@@ -214,15 +214,30 @@ function formatResponsesMessages(messages: Message[]): ResponseInputItem[] {
...
-        content: contentItems,
-      } as unknown as ResponseInputItem)
+      if (role === 'assistant') {
+        // Only the string EasyInputMessage form is a valid assistant *input* shape
+        // without the full ResponseOutputMessage metadata (id/status/annotations),
+        // which has already been discarded by this point. Non-text assistant history
+        // has no valid input shape either. Replaying verbatim output items is
+        // tracked in https://github.com/strands-agents/harness-sdk/issues/3389.
+        if (contentItems.some((item) => item.type !== 'output_text')) {
+          logger.warn('non-text assistant history content has no valid responses api input shape | dropping')
```
[the remark this answers is not in the document]
Comment (agent-of-mkmeral) [reply strands-ts/src/models/openai/responses-adapter.ts]: Fixed in f4c21a48 — the warning now interpolates the dropped content type(s): `content_type=<input_image>` etc., matching the `block_type=<...>`/`content_type=<...>` convention used elsewhere in this file. Python side updated identically for parity.

File: strands-ts/src/models/openai/responses-adapter.ts [hunk 2]
```diff
@@ -171,7 +171,7 @@ function formatResponsesMessages(messages: Message[]): ResponseInputItem[] {
         case 'citationsBlock': {
           const citBlock = block as { content: Array<{ text: string }> }
           for (const c of citBlock.content) {
-            contentItems.push({ type: 'output_text', text: c.text })
+            contentItems.push({ type: role === 'user' ? 'input_text' : 'output_text', text: c.text })
```
[the remark this answers is not in the document]
Comment (agent-of-mkmeral) [reply strands-ts/src/models/openai/responses-adapter.ts:174]: Added in f4c21a48 — `serializes user-role citations content as input_text` now exercises the user-role branch and asserts the full `{ role: 'user', content: [{ type: 'input_text', ... }] }` shape.

File: strands-py/src/strands/models/openai_responses.py [hunk 3]
```diff
@@ -634,12 +634,25 @@ def _format_request_messages(cls, messages: Messages) -> list[dict[str, Any]]:
...
-                    }
-                )
+                if role == "assistant":
+                    # Only the string EasyInputMessage form is a valid assistant *input* shape
+                    # without the full output-item metadata (id/status/annotations) that has
+                    # already been discarded by this point. Replaying verbatim output items is
+                    # tracked in https://github.com/strands-agents/harness-sdk/issues/3389.
+                    if any(part.get("type") != "output_text" for part in formatted_contents):
+                        logger.warning(
+                            "non-text assistant history content has no valid Responses API input shape | dropping"
```
[the remark this answers is not in the document]
Comment (agent-of-mkmeral) [reply strands-py/src/strands/models/openai_responses.py]: Fixed in f4c21a48 — the warning now reports the dropped type(s) as `content_type=<%s>` per the module convention, and the TS counterpart was updated identically. Also added a test for an assistant turn with *only* non-text content (drops entirely), which covers the remaining codecov partial branch.

**Comment by agent-of-mkmeral** [comment]
Pushed **f4c21a48** addressing the codecov failure and all review comments:

**Codecov (`codecov/patch` 85.71% → 100% of diff):** the missed partial was the `if text:` false branch in `_format_request_messages` (an assistant turn whose content is *entirely* non-text collapses to nothing). Added `test_format_request_messages_assistant_only_non_text_content_dropped_entirely` covering it — locally the diff range now has no missing lines/branches (`--cov-branch` shows remaining partials only in pre-existing stream-handling code).

**Review comments:**
1. **Drop warnings now identify what was dropped** (both SDKs): `content_type=<input_image> | non-text assistant history content has no valid Responses API input shape | dropping` — matching the module convention.
2. **User-role `citationsBlock` branch now tested** (TS): asserts `input_text` is emitted.
3. **Edge case (assistant turn with only non-text content disappearing from history):** now pinned by explicit tests in both SDKs documenting the behavior. Agreed it's a role-alternation risk on strict backends in theory, but fabricating an empty assistant string seems worse; proper verbatim replay is #3389's scope.

**Verification:** Py 105/105 pass, ruff check+format clean · TS 52/52 in the responses suite, full unit-node suite green (the only failures were my runner's OTEL env vars polluting telemetry tests — pass with clean env), `tsc` (src project) clean, eslint + prettier clean.

**Review by opieter-aws — APPROVED** [review]

**Review by yonib05 — APPROVED** [review]

**Comment by agent-of-mkmeral** [comment]
Follow-up on the CI status at `f4c21a48` — everything is now green:

- **codecov/patch**: ✅ SUCCESS — "All modified and coverable lines are covered by tests" (the earlier 85.71% partial-branch gap was closed by the new entirely-non-text assistant turn test).
- **Python: Integration Test**: ✅ SUCCESS
- **TypeScript: Secure Integration Test**: ✅ SUCCESS — and crucially, the live Mantle test this PR exists to fix passed with team credentials:
  ```
  ✓ test/integ/models/openai/mantle.test.node.ts (4 tests) 9639ms
    ✓ reaches Mantle via bedrockMantleConfig on the Chat Completions API
    ✓ reaches Mantle via bedrockMantleConfig on the Responses API
    ✓ supports server-side stateful conversations
    ✓ handles reasoning content across multi-turn conversations  ← the #3388 repro
  ```
  This is the confirmation the PR description asked for (I couldn't run it locally due to `bedrock-mantle:CreateInference` being denied on my credentials).
- **CI Gate**: ✅ SUCCESS — all unit tests (Py 3.10–3.14 linux/windows, Node 20/22/24 across 3 OSes), lint, type-check, docs, pack-install all pass.
- **Review comments**: all 3 threads addressed in `f4c21a48` and confirmed resolved by the reviewer's re-review (Approve ✅, "No outstanding concerns").

The only remaining red ✗ is **`2 Approvers for Bots`** — that's the bot-PR review gate requiring 2 maintainer approvals; it isn't a test failure and will clear once two maintainers approve. Ready for maintainer review.

File: strands-ts/src/models/openai/responses-adapter.ts [hunk 4]
```diff
@@ -214,15 +214,33 @@ function formatResponsesMessages(messages: Message[]): ResponseInputItem[] {
...
-        role,
-        content: contentItems,
-      } as unknown as ResponseInputItem)
+      if (role === 'assistant') {
+        // Only the string EasyInputMessage form is a valid assistant *input* shape
+        // without the full ResponseOutputMessage metadata (id/status/annotations),
+        // which has already been discarded by this point. Non-text assistant history
+        // has no valid input shape either. Replaying verbatim output items is
+        // tracked in https://github.com/strands-agents/harness-sdk/issues/3389.
+        const droppedTypes = contentItems.filter((item) => item.type !== 'output_text').map((item) => item.type)
```
Comment (pgrayy) [inline strands-ts/src/models/openai/responses-adapter.ts:224]: Do we know what kind of content types we risk dropping?

**Review by pgrayy — APPROVED** [review]

File: strands-ts/src/models/openai/responses-adapter.ts (hunk 4 above)
Comment (opieter-aws) [reply strands-ts/src/models/openai/responses-adapter.ts:224]: The set is bounded to `input_image` and `input_file`.

`contentItems` only receives non-`output_text` entries from `imageBlock` and `documentBlock`, everything else either becomes `output_text `(so it’s kept, not dropped) or never enters `contentItems` at all
